File: src/Server/Context.cs
namespace Belin.UI.UI;

using System.Text.Json.Serialization;

/// <summary>
/// Defines contextual modifiers.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum Context {

	/// <summary>
	/// A danger.
	/// </summary>
	Danger,

	/// <summary>
	/// A warning.
	/// </summary>
	Warning,

	/// <summary>
	/// An information.
	/// </summary>
	Info,

	/// <summary>
	/// A success.
	/// </summary>
	Success
}

/// <summary>
/// Provides extension members for contextual modifiers.
/// </summary>
public static class ContextExtensions {
	extension(Context context) {

		/// <summary>
		/// The CSS class of this context.
		/// </summary>
		public string CssClass => context.ToString().ToLowerInvariant();

		/// <summary>
		/// The icon corresponding to this context.
		/// </summary>
		public string Icon => context switch {
			Context.Danger => "error",
			Context.Warning => "warning",
			Context.Info => "info",
			Context.Success => "check_circle"
		};
	}
}

File: Sources/Client/Context.ts
/**
 * Defines contextual modifiers.
 */
export const Context = Object.freeze({

	/**
	 * A danger.
	 */
	Danger: "Danger",

	/**
	 * A warning.
	 */
	Warning: "Warning",

	/**
	 * An information.
	 */
	Info: "Info",

	/**
	 * A success.
	 */
	Success: "Success"
});

/**
 * Defines contextual modifiers.
 */
export type Context = typeof Context[keyof typeof Context];

/**
 * Returns the CSS class of the specified context.
 * @param context The context.
 * @returns The CSS class of the specified context.
 */
export function cssClass(context: Context): string {
	return context.toLowerCase();
}

/**
 * Gets the icon corresponding to the specified context.
 * @param context The context.
 * @returns The icon corresponding to the specified context.
 */
export function icon(context: Context): string {
	switch (context) {
		case Context.Danger: return "error";
		case Context.Success: return "check_circle";
		case Context.Warning: return "warning";
		default: return "info";
	}
}
